[APM] Migrate remaining tx-based visualisations

dgieselaar opened the pull request:
Closes #152693

dgieselaar committed:
[APM] Migrate tx throughput/failure rate charts

sqren commented on x-pack/plugins/apm/server/lib/transaction_groups/get_failed_transaction_rate.ts:
```diff
@@ -40,23 +36,25 @@ export async function getFailedTransactionRate({
   end: number;
-  numBuckets?: number;
   offset?: string;
+  documentType: ApmServiceTransactionDocumentType;
+  rollupInterval: RollupInterval;
+  bucketSizeInSeconds: number;
```
```suggestion
  documentType: ApmServiceTransactionDocumentType;
  rollupInterval: RollupInterval;
  bucketSizeInSeconds: number;
```

sqren commented on x-pack/plugins/apm/server/routes/services/route.ts:
```diff
@@ -546,23 +552,21 @@ const serviceThroughputRoute = createApmServerRoute({
       offset,
       start,
       end,
+      documentType,
+      rollupInterval,
```
Perhaps co-locate these. Does it ever make sense to pass one and not the other? Feels like they belong together.
```suggestion
      apmDataSource: { documentType, rollupInterval }
```

dgieselaar replied:
Query parameters need to be flat, and in some cases `bucketSizeInSeconds` should be added as well. The other option is "merge" them in the route, but it will complicate the type a little. E.g. you have to write `ApmDataSource<ApmServiceTransactionDocumentType>`. Not sure if it's worth it but I can make that change if you feel strongly about it.

sqren commented on x-pack/test/apm_api_integration/tests/error_rate/service_apis.spec.ts:
```diff
@@ -67,6 +67,16 @@ export default function ApiTest({ getService }: FtrProviderContext) {
             ...commonQuery,
             kuery: `processor.event : "${processorEvent}"`,
```
Do we still need this, now we are able to specify the `documentType`?

dgieselaar replied:
`GET /internal/apm/services/{serviceName}/service_overview_instances/main_statistics` doesn't use `documentType` (yet).

dgieselaar commented:
Putting this back in draft, I pushed a commit with temporary changes to test out some performance improvements.

gbamparop commented on x-pack/plugins/apm/public/components/shared/charts/failed_transaction_rate_chart/index.tsx:
```diff
@@ -71,6 +73,14 @@ export function FailedTransactionRateChart({
+    start,
+    end,
+    numBuckets: 100,
+    kuery: '',
```
Is this missing on purpose here?

dgieselaar replied:
no, good catch - must have copied it from one of the alert details charts. thanks!

dgieselaar commented on x-pack/plugins/apm/public/components/shared/charts/failed_transaction_rate_chart/index.tsx:
```diff
@@ -71,6 +73,14 @@ export function FailedTransactionRateChart({
+    start,
+    end,
+    numBuckets: 100,
+    kuery: '',
```
```suggestion
    kuery,
```

dgieselaar committed:
Update x-pack/plugins/apm/public/components/shared/charts/failed_transaction_rate_chart/index.tsx

dgieselaar commented on x-pack/plugins/apm/public/components/alerting/ui_components/alert_details_app_section/throughput_chart.tsx:
```diff
@@ -72,13 +83,16 @@ function ThroughputChart({
                   end,
                   transactionType,
                   transactionName: undefined,
```
Yes, this is for the alert details section, alerts don't have a transaction name.

kpatticha replied:
what do you think of changing the `transactionName`  to be required https://github.com/elastic/kibana/blob/170e5013c430d357fe6fe33900ab226f5021c468/x-pack/plugins/apm/server/routes/services/route.ts#L526-L528

It doesn't have to be part of this PR though

kpatticha reviewed (approved):
Overall looks good to me.

Only the failed transaction rate seems a bit off 

![image](https://user-images.githubusercontent.com/3369346/228516053-c5e8264c-4839-449a-9530-4f5c0c981d2a.png)

could you please check it 🙏

dgieselaar committed:
Remove event.outcome filter and add API tests for service tx metrics

dgieselaar committed:
Merge branch 'service-overview-rollups-throughput-failure-rate' of github.com:dgieselaar/kibana into service-overview-rollups-throughput-failure-rate

dgieselaar committed:
Merge branch 'main' of github.com:elastic/kibana into service-overview-rollups-throughput-failure-rate

dgieselaar commented:
@kpatticha grazie! good catch. There was an event.outcome term filter for failure rate, but the field doesn't exist on service tx metrics, rather success/failure counts are tracked on the same document. I've removed the filter and added an API test for service tx metrics which we didn't have before.

dgieselaar committed:
Fix type errors
